agent/engine/engine_windows_integ_test.go

```diff
@@ -506,3 +516,370 @@ func verifyContainerCredentialSpec(client *sdkClient.Client, id, credentialspecO
+}
+
+// TestManagedAgentEvent validates the emitted container events for a started and a stopped managed agent.
+func TestManagedAgentEvent(t *testing.T) {
+	// the execcmd feature is not supported for Win2016
+	if windows2016, _ := config.IsWindows2016(); windows2016 {
+		return
+	}
+	testcases := []struct {
+		Name                 string
+		ExpectedStatus       apicontainerstatus.ManagedAgentStatus
+		ManagedAgentLifetime time.Duration
+		ShouldBeRunning      bool
+	}{
+		{
+			Name:                 "Confirmed emit RUNNING event",
+			ExpectedStatus:       apicontainerstatus.ManagedAgentRunning,
+			ManagedAgentLifetime: 1,
+			ShouldBeRunning:      true,
+		},
+		{
+			Name:                 "Confirmed emit STOPPED event",
+			ExpectedStatus:       apicontainerstatus.ManagedAgentStopped,
+			ManagedAgentLifetime: 0,
+			ShouldBeRunning:      false,
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.Name, func(t *testing.T) {
+
+			const (
+				testTaskId        = "exec-command-agent-test-task"
+				testContainerName = "exec-command-agent-test-container"
+			)
+
+			client, err := sdkClient.NewClientWithOpts(sdkClient.WithHost(endpoint), sdkClient.WithVersion(sdkclientfactory.GetDefaultVersion().String()))
+			require.NoError(t, err, "Creating go docker client failed")
+
+			testExecCmdHostBinDir := "C:\\Program Files\\Amazon\\ECS\\managed-agents\\execute-command\\bin"
+
+			taskEngine, done, _ := setupEngineForExecCommandAgent(t, testExecCmdHostBinDir)
+			defer done()
+
+			testTask := createTestExecCommandAgentTask(testTaskId, testContainerName, time.Minute*tc.ManagedAgentLifetime)
+			execAgentLogPath := filepath.Join("C:\\ProgramData\\Amazon\\ECS\\exec", testTaskId)
+			err = os.MkdirAll(execAgentLogPath, 0644)
+			require.NoError(t, err, "error creating execAgent log file")
+			_, err = os.Stat(execAgentLogPath)
+			require.NoError(t, err, "execAgent log dir doesn't exist")
+			err = os.MkdirAll(execcmd.ECSAgentExecConfigDir, 0644)
+			require.NoError(t, err, "error creating execAgent config dir")
+
+			go taskEngine.AddTask(testTask)
+
+			verifyContainerRunningStateChange(t, taskEngine)
+			verifyTaskRunningStateChange(t, taskEngine)
+
+			if tc.ShouldBeRunning {
+				containerMap, _ := taskEngine.(*DockerTaskEngine).state.ContainerMapByArn(testTask.Arn)
+				cid := containerMap[testTask.Containers[0].Name].DockerID
+				verifyMockExecCommandAgentIsRunning(t, client, cid)
+			}
+			waitDone := make(chan struct{})
+
+			go verifyExecAgentStateChange(t, taskEngine, tc.ExpectedStatus, waitDone)
+
+			timeout := false
+			select {
+			case <-waitDone:
+			case <-time.After(20 * time.Second):
+				timeout = true
+			}
+			assert.False(t, timeout)
+
+			if tc.ShouldBeRunning {
+				containerMap, _ := taskEngine.(*DockerTaskEngine).state.ContainerMapByArn(testTask.Arn)
+				cid := containerMap[testTask.Containers[0].Name].DockerID
+				// Kill the existing container now
+				err = client.ContainerKill(context.TODO(), cid, "SIGKILL")
```

Comment:
Will updating the desired status of the task to "Stopped" work here?

Reply:
yes, does it matter whether it is killed or stopped? As of right now the Linux version of this test does neither. Im not sure how it guarantees the container is stopped

Reply:
Might be worth trying something like this: https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/docker_task_engine_test.go#L464-L471

Accessing the task internal state is racy

Reply:
Could I stop the task like how Linux does it? 
https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/engine_sudo_linux_integ_test.go#L457-L469

Reply:
yes, that works